---
title: "Preprocessing GWAS-derived data"
author: "Fabricio Almeida-Silva"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE,
                      warning=FALSE, 
                      message=FALSE,
                      tidy.opts=list(width.cutoff=60),
                      tidy=TRUE)
```

## Data description

- _snp_list.csv_: List of SNPs related to resistance to biotic stress along with reference to their original articles and the pathogen/pest to which the SNPs are related. This database was created by manually curating the scientific literature on genome-wide association studies that aimed at identifying loci affecting disease/pest resistance. The search strategy used on Google Scholar was: _intitle:"soybean" AND "GWAS" AND "resistance"_. Last update: March 2021.

## Data pre-processing and exploratory analyses

First, we must load the data and filter data to keep only SNPs related to resistance to insects.

```{r}
library(BioNERO)
library(here)
library(tidyverse)
species <- c("Slitura", "Aglycines", "Hglycines")
snplist <- read.csv(here("data", "snp_list.csv"), header=TRUE, sep=";") %>%
  dplyr::select(!c(1:3)) %>%
  distinct() %>%
  filter(Organism %in% species) %>%
  mutate(SNP = str_trim(SNP)) %>%
  arrange(Organism, SNP) %>%
  distinct(SNP, .keep_all = TRUE)
```

Now, let's see if we have SNPs from different genome versions, or different marker types.

```{r}
table(snplist$MarkerType)
table(snplist$GenomeVersion)
```

## Getting genome positions for the SNPs

The .vcf files containing SNP positions for *Wm82.a2.v1* and *Wm82.a1.v1* were downloaded from Soybase.

Load the files with SNP positions.

```{r}
positions_a2 <- read.csv(here("data", "snp_positions_a2.v1.txt"), 
                         header=T, sep = "\t")
positions_a1 <- read.csv(here("data", "snp_positions_a2.v1.txt"), 
                         header=T, sep = "\t")
names(positions_a2) <- c("chr", "pos", "id")
names(positions_a1) <- c("chr", "pos", "id")
head(positions_a2)
head(positions_a1)
```

Replace BARC IDs with dbSNP IDs:

```{r barc_to_dbsnp}
snplist[startsWith(snplist$SNP, "BARC"), ]

# IDs were retrieved by manual searching on Soybase
snplist$SNP <- gsub("BARC-012259-01773", "Chr01-2195303", snplist$SNP)
snplist$SNP <- gsub("BARC-012295-01800", "ss715578714", snplist$SNP)
snplist$SNP <- gsub("BARC-019001-03050", "Chr01-17817555", snplist$SNP)
snplist$SNP <- gsub("BARC-047665-10370", "ss715583614", snplist$SNP)
snplist$SNP <- gsub("BARC-048271-10520", "Chr01-812542", snplist$SNP)
snplist$SNP <- gsub("BARC-G01477-00243", "ss715581136", snplist$SNP)
```

Create new columns containing chromosome and position for the SNPs.

```{r}
snp_pos <- merge(snplist, positions_a2, 
                 by.x="SNP", by.y="id", all.x=TRUE)
```

Some SNPs do not have IDs. For these cases, SNP names are identified as `ChrX-pos`. Let's include genomic coordinates for these SNPs.

```{r}
source(here::here("code", "utils.R"))

snp_pos <- add_coordinates(snp_pos)
snp_pos$pos <- NULL

# Check if they have all been replaced
nrow(snp_pos[rowSums(is.na(snp_pos)) > 0, ])
```

Fix chromosome names (Chr1-Chr9 will be replaced by Chr01-Chr09):

```{r}
snp_pos <- chr_rename(snp_pos)
```

SNP positions on a1.v1 could not be matched to SNPs in a2.v1. Then, we will remove them. Finally, we will tidy the data for further analyses.

```{r save_snp_granges}
library(GenomicRanges)
snp_ranges <- snp_pos %>%
  filter(GenomeVersion != "a1.v1") %>%
  filter(!startsWith(chr, "scaffold")) %>%
  dplyr::select(chr, start, end, Organism, Trait, SNP, DOI)

# GRanges
snp_granges <- makeGRangesFromDataFrame(snp_ranges)
snp_granges$Organism <- snp_ranges$Organism
snp_granges$Trait <- snp_ranges$Trait
snp_granges$SNP <- snp_ranges$SNP
snp_granges$DOI <- snp_ranges$DOI
```

To wrap it up, making a table of studies included in this study. This will be Table 1 in the manuscript.

```{r table1}
table1 <- as.data.frame(snp_granges)
t1 <- table1 %>%
  dplyr::count(DOI, Organism) %>%
  arrange(DOI, Organism)
write_tsv(t1, file = here::here("products", "tables", "main_table1.tsv"))
```

Save data for further use.

```{r}
# Remove redundant entries (same SNP identified by different papers)
snp_ranges <- snp_ranges[!duplicated(snp_ranges$SNP), ]

# GRanges
snp_granges <- makeGRangesFromDataFrame(snp_ranges)
snp_granges$Organism <- snp_ranges$Organism
snp_granges$Trait <- snp_ranges$Trait
snp_granges$SNP <- snp_ranges$SNP
snp_granges$DOI <- snp_ranges$DOI


# GRangesList
snp_grangeslist <- split(snp_granges, snp_granges$Organism)

# Saving data
save(snp_granges, snp_grangeslist, 
     file = here("products", "result_files", "snp_granges.rda"),
     compress = "xz")
```


## Session information

This document was created under the following conditions:

```{r}
sessionInfo()
```